docs/source/howto/aforms.rst

```diff
@@ -0,0 +1,56 @@
+.. _aforms:
+
+How to extract AcroForm fields from a PDF using Python and PDFMiner
+********************************
+
+Before you start, make sure you have :ref:`installed pdfminer.six<install>`.
+
+The second thing you need is a PDF with AcroForms (as found in PDF files with fillable forms or multiple choices).
+
+It won't work for XFA forms.
+
+.. code-block:: python
+
+    from pdfminer.pdfparser import PDFParser
+    from pdfminer.pdfdocument import PDFDocument
+    from pdfminer.pdftypes import resolve1, PDFObjRef
+    from pdfminer.psparser import PSLiteral, PSKeyword
+    from pdfminer import utils
+    
+    data = {}
+    
+    with open(file_path, 'rb') as fp:
+        parser = PDFParser(fp)
+        
+        doc = PDFDocument(parser)
+        res = resolve1(doc.catalog)
+
+        if 'AcroForm' in res:
+            fields = resolve1(doc.catalog['AcroForm'])['Fields']
+
+            for f in fields:
+                field = resolve1(f)
+                name, value = field.get('T'), field.get('V')
+
+                # decode name
+                name = name.decode()
+
+                if value:
+
+                    # resolve indirect obj
+                    while isinstance(value, PDFObjRef):
+                        value = resolve1(value)
+
+                    # decode PSLiteral, PSKeyword
+                    if isinstance(value, (PSLiteral, PSKeyword)):
+                        value = value.name
+
+                    # decode bytes
+                    if isinstance(value, bytes):
+                        value = utils.decode_text(value)
+              
+                data.update({name: value})    
```

Comment:
I think it makes sense to extract a method that has the `PDFDocument` as an input and returns the `data` object. Then we give the reader a piece of code that they can easily reuse / map over multiple documents.

Reply:
Oh, makes sense, I'll see about it.

Question!

You may remember I was lamenting pdfminer was slow when creating the `PDFDocument`.
I found the reason: basically when creating `PDFDocument`all of the ObjRef are being resolved.

For extracting the AcroForm fields this is overkill, at least for my forms (where there are lots of unrelated ObjRef).
It should be better to get the catalog, check if there's an `AcroForm` key and if there is just resolve its ObjRef.

Does this make sense? I fear this approach would need some major work...

Maybe I'll try a custom class in the weekend to see how much work it is.

Reply:
Method extracted.

Reply:
I just ran the test suite and noticed that it took 40 seconds to parse _nonfree/nlp2004slides.pdf_. Maybe we can use that to see if we can improve the speed. Different issue / PR though ;)